### PR TITLE
WIP: initialize repo

### PR DESCRIPTION
**Pull request by robertsLando**

### CODE REVIEW

**Comment by robertsLando** [comment]
@mcollina Hi Matteo, this is just a draft but could you take a quick look at it and tell me if the approach I'm using it's correct? :)

Any suggestion is welome, tomorrow I will work on persistences (they will only be available using config file

File: lib/authorizer.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
+'use strict'
+
+var hasher = require('pbkdf2-password')()
```
Comment (mcollina) [inline lib/authorizer.js:3]: Maybe there is something better than this around, how about `secure-password`? Or some other algorithm that is in Node core?

File: lib/authorizer.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    that._authenticate(client, user, pass, cb)
+  }
+})
+
+/**
+ * It returns the authorizePublish function to plug into mosca.Server.
+ *
+ * @api public
+ */
+Authorizer.prototype.__defineGetter__('authorizePublish', function () {
```
Comment (mcollina) [inline lib/authorizer.js:37]: Please modernize this code from Mosca :)

File: lib/cli.js [hunk 3]
```diff
@@ -0,0 +1,330 @@
+'use strict'
+
+var pkg = require('../package')
+var commander = require('commander')
```
Comment (mcollina) [inline lib/cli.js:4]: I think we should use yargs for this instead.

File: lib/cli.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        authorizer.users = JSON.parse(data)
+        cb(null, authorizer)
+      } catch (err) {
+        cb(err)
+      }
+    })
+  } else {
+    cb(null, null)
+  }
+}
```
Comment (mcollina) [inline lib/cli.js:57]: I would use async/await in the whole of this codebase. It'd be significantly easier to read and there is no need for perf.

Comment (robertsLando) [reply lib/cli.js:54]: I thought you was a `callback` lover so I made all the codebase using them :)

File: lib/cli.js (hunk 3 above)
Comment (robertsLando) [reply lib/cli.js]: Is there any pro/cons? I see they have same stats and commander is lighter then yargs in terms of size

File: lib/authorizer.js (hunk 1 above)
Comment (robertsLando) [reply lib/authorizer.js:3]: I only used `bcrypt` for this kind of things, you are more expert then me in such things.

Could we use [crypto](https://nodejs.org/api/crypto.html) ?

Comment (robertsLando) [reply lib/authorizer.js]: Check out `lib/hasher.js` and tell me if it's ok for you :)

File: lib/cli.js (hunk 4 above)
Comment (robertsLando) [reply lib/cli.js:54]: Is there something I can keep using async/await?

**Comment by robertsLando** [comment]
I will start with tests once I have defined the structure of the project, just to don't waste time rewriting them more times :)

File: lib/authorizer.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  var { salt, hash } = await hasher.generateHashPassword(pass.toString())
+
+  that.users[user] = {
+    salt: salt,
+    hash: hash,
+    authorizePublish: authorizePublish,
+    authorizeSubscribe: authorizeSubscribe
+  }
+
+  return this
```
Comment (mcollina) [inline lib/authorizer.js:113]: I would not return this.

File: lib/cli.js [hunk 6]
```diff
@@ -0,0 +1,330 @@
+'use strict'
+
+const pkg = require('../package')
+const commander = require('commander')
+const path = require('path')
+const Authorizer = require('./authorizer')
+var { readFile, writeFile } = require('fs')
```
Comment (mcollina) [inline lib/cli.js:7]: You can possibly use `fs.promises`.

File: lib/authorizer.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ * mosca.Authorizer's responsibility is to give an implementation
+ * of mosca.Server callback of authorizations, against a JSON file.
+ *
+ * @param {Object} users The user hash, as created by this class
+ *  (optional)
+ * @api public
+ */
+function Authorizer (users) {
+  this.users = users || {}
```
Comment (mcollina) [inline lib/authorizer.js:16]: I would recommend using a map instead, otherwise it's possible to have prototype pollution attacks.

File: lib/authorizer.js (hunk 5 above)
Comment (robertsLando) [reply lib/authorizer.js]: Was there from Mosca, removed :)

File: lib/cli.js (hunk 6 above)
Comment (robertsLando) [reply lib/cli.js:7]: Nodejs support? 10+?

**Comment by robertsLando** [comment]
@mcollina What test plugin do you suggest to use for tests?

File: lib/cli.js (hunk 6 above)
Comment (mcollina) [reply lib/cli.js:7]: aedes is 10+, so yes

Comment (robertsLando) [reply lib/cli.js]: I mean the support for `fs.promises`

**Comment by robertsLando** [comment]
@mcollina should be almost ready now :)

File: lib/cli.js [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+
+  await setupAuthorizer()
+
+  // STATS ------------
+
+  // if (config.disableStats) {
+  //   //disable stats
+  // }
+
+  // SERVERS ------------
```
Comment (mcollina) [inline lib/cli.js]: why is this commented out?

File: lib/cli.js [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+      .option('--maxClientsIdLength <n>', 'broker option to override MQTT 3.1.0 clients Id length limit', parseInt)
+      .option('--heartbeatInterval <n>', 'interval in millisconds at which broker beats its health signal in $SYS/<broker.id>/heartbeat', parseInt)
+      .option('--connectTimeout <n>', 'maximum waiting time in milliseconds waiting for a CONNECT packet.', parseInt)
+      .option('--key <file>', "the server's private key")
+      .option('--cert <file>', 'the certificate issued to the server')
+      .option('--reject-unauthorized', 'reject clients using self signed certificates', true)
+      .option('--tls-port <n>', 'the TLS port to listen to', parseInt)
+      .option('--ws-port <n>', 'start an mqtt-over-websocket server on the specified port', parseInt)
+      .option('--wss-port <n>', 'start an mqtt-over-secure-websocket server on the specified port', parseInt)
+    //  .option('--disable-stats', 'disable the publishing of stats under $SYS', true)
```
Comment (mcollina) [inline lib/cli.js]: ??

File: test/config/redisConfig.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  //   options: {
+  //     url: 'mongodb://127.0.0.1/aedes'
+  //   }
+  // },
+  // mq: {
+  //   name: 'mongodb',
+  //   options: {
+  //     url: 'mongodb://127.0.0.1/aedes'
+  //   }
+  // },
```
Comment (mcollina) [inline test/config/redisConfig.js:34]: please remove.

File: test/helper.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+      else resolve()
+    })
+  })
+}
+
+async function stop (setup) {
+  await close(setup.broker)
+
+  setup.broker.persistence.destroy()
+  // setup.broker.mq.close()
```
Comment (mcollina) [inline test/helper.js:28]: why is this needed?

Comment (robertsLando) [reply test/helper.js]: If you mean `setup.broker.persistence.destroy()` it is needed because without that test never ends. If you mean the commented line, I have just removed it

File: lib/cli.js (hunk 9 above)
Comment (robertsLando) [reply lib/cli.js]: check latest version

File: lib/cli.js (hunk 8 above)
Comment (robertsLando) [reply lib/cli.js]: check latest version

File: config.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  //     }
+  //   },
+  //   persistence: {
+  //     name: 'redis',
+  //     options: {}
+  //   },
+  //   mq: {
+  //     name: 'redis',
+  //     options: {}
+  //   },
```
[the remark this answers is not in the document]
Comment (robertsLando) [reply config.js:42]: Them are there just for reference in case users need an example with persistences. Now I remove it

**Comment by robertsLando** [comment]
@mcollina I have fixed last thing, let me know if it's ok now

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        node-version: [10.x, 12.x, 14.x]
+    services:
+      redis:
+        image: redis
+        ports:
+        - 6379:6379
+      mongodb:
+        image: mongo:3.4.23
+        ports:
+          - 27017:27017
```
Comment (mcollina) [inline .github/workflows/ci.yml:28]: I would prefer if we had separate runs for Redis and Mongo. Running all on the same workflow is going to cause similar problems we had in Mosca.

Comment (robertsLando) [reply .github/workflows/ci.yml:28]: How should I manage tests so? Should I use env vars for the persistence to test?

Comment (mcollina) [reply .github/workflows/ci.yml:28]: I would do that, yes.

**Comment by robertsLando** [comment]
@mcollina tell me if it's ok now :smile:
